docs/design/trimmed-assemblies.md

```diff
@@ -0,0 +1,183 @@
+
+## .NET 6
+
+### `IsTrimmable` `AssemblyMetadataAttribute`
+
+An assembly-level `AssemblyMetadataAttribute` may be placed on an assembly to indicate that it should be trimmed:
+
+```csharp
+[AssemblyMetadataAttribute("IsTrimmable", "true")]
+```
+
+The behavior is the same as the `IsTrimmable` MSBuild metadata, so that:
+- Assemblies with this attribute are trimmed with the global `TrimMode`
+- Assemblies without this attribute are rooted, and given the `copy` action
+
+The only understood value is `true`. Adding `[AssemblyMetadataAttribute("IsTrimmable", "false")]` will have no effect on the linker's behavior, because unattributed assemblies are assumed not to be trimmable by default. We will issue a warning in this case, to discourage misleading use of the attribute.
+
+If `IsTrimmable` MSBuild metadata is set for an assembly, this overrides the `IsTrimmable` attribute. This allows a developer to opt an assembly into trimming even if it does not have the attribute, or to disable trimming of an assembly that has the attribute.
+
+Instead of using `IsTrimmable` metadata in the SDK to trim netcoreapp by default, we will ship netcoreapp assemblies and OOB assemblies that are conceptually part of the BCL with `IsTrimmable` `AssemblyMetadataAttribute`, and not set `IsTrimmable` MSBuild metadata in the SDK. The MSBuild metadata will now be used only as a developer opt-in or opt-out.
+
+### `TrimmableAssembly`
+
+This ItemGroup contains assembly names that will be opted into trimming via `IsTrimmable` metadata. For simple cases, this provides an easier way to enable trimming of additional assemblies, without requiring a custom MSBuild target.
+
+```xml
+<ItemGroup>
+  <TrimmableAssembly Include="MyAssembly" />
+</ItemGroup>
+```
+
+The above will opt `MyAssembly.dll` into trimming. Note that the ItemGroup should contain assembly names without an extension, similar to [`TrimmerRootAssembly`](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming-options#root-assemblies). Before .NET 6 this would have been done with a target:
+
+```xml
+<Target Name="ConfigureTrimming"
+        BeforeTargets="PrepareForILLink">
+  <ItemGroup>
+    <ManagedAssemblyToLink Condition="'%(Filename)' == 'MyAssembly'">
+      <IsTrimmable>true</IsTrimmable>
+    </ManagedAssemblyToLink>
+  </ItemGroup>
+</Target>
+```
+
+### `TrimAllAssemblies`
+
+This property simplifies the process of enabling trimming for all assemblies. It is equivalent to setting `IsTrimmable` to `true` on every assembly that is input to the linker. For example:
+
+```xml
+<PropertyGroup>
+  <TrimAllAssemblies>true</TrimAllAssemblies>
+</PropertyGroup>
+```
+
+Before .NET 6 this would have been done with a target:
+
+```xml
+<Target Name="ConfigureTrimming"
+        BeforeTargets="PrepareForILLink">
+  <ItemGroup>
+    <ManagedAssemblyToLink>
+      <IsTrimmable>true</IsTrimmable>
+    </ManagedAssemblyToLink>
+  </ItemGroup>
+</Target>
+```
+
+## Alternatives considered
```

Comment:
I think we should add a section considering future evolution.
* Very long term we want SDKs to enable full trimming on everything.
  * This would require opt-out mechanisms - we would likely have to support `[AssemblyMetadataAttribute("IsTrimmable", "false")] - or maybe even support specifying link action in the attribute.
* Medium term we may want to "upgrade" the console app trimming default from `copyused` to `link`.
  * This would require per-assembly opt-out - so some ability (in the SDK) to specify link action per-assembly (`TrimMode`)

I think we're fine on all of these, but we should mention these and how we plan to go about them.

Reply:
Thanks - I added a Future section. PTAL.